Unicode awareness Part 2 -- copytext()

MrPerson opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request


This is the transition of all copytext() calls to be unicode aware and also some nearby calls in the same functions. Mostly this is just replacing copytext() with copytext_char() as a terrible character limiter with all the same issues that were pointed out in part 1.

I replaced a ton of
````
var/something = sanitize(input())
something = copytext(something, 1, MAX_MESSAGE_LEN)
````

with a single stripped_input() call. stripped_input() calls html_encode(), trim(), and some other sanitization so there shouldn't be any major issues there.

This is still VERY rough btw; DNA is a mess, the status displays are complete ass, there's a copytext() in code\datums\shuttles.dm that I'm not sure what to do with, ~there's an issue with stuttering I need to figure out~, and I didn't touch anything in the tools folder.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Not chopping unicode chars in half.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

MrPerson committed:
Unicode support Part 2 -- copytext()

This is the transition of all copytext() calls to be unicode aware and also some nearby calls in the same functions. Most things are just replacing copytext() with copytext_char() as a terrible character limiter but a few others were slightly more involved.

I replaced a ton of
````
var/something = sanitize(input())
something = copytext(something, 1, MAX_MESSAGE_LEN)
````

with a single stripped_input() call. stripped_input() already calls html_encode(), trim(), and some other sanitization so there shouldn't be any major issues there.

This is still VERY rough btw; DNA is a mess, the status displays are complete ass, there's a copytext() in code\datums\shuttles.dm that I'm not sure what to do with, and I didn't touch anything in the tools folder. I haven't tested this much at all yet, I only got it to compile earlier this morning. There's also likely to be weird bugs until I get around to fixing length(), findtext(), and the rest of the string procs.

MrPerson committed:
Makes the code functional

MrPerson commented:
I've marked this DNM and made it a draft instead of a real PR because it's gonna fail travis until the first part is merged.

AnturK commented on code/__DEFINES/typeids.dm:
```diff
@@ -2,7 +2,7 @@
 #define TYPEID_NULL "0"
 #define TYPEID_NORMAL_LIST "f"
 //helper macros
-#define GET_TYPEID(ref) ( ( (length(ref) <= 10) ? "TYPEID_NULL" : copytext(ref, 4, length(ref)-6) ) )
```
I think it's safe to assume there's no special characters in refs.

AnturK commented on code/__HELPERS/sanitize_values.dm:
```diff
@@ -74,7 +74,7 @@
 
 /proc/sanitize_ooccolor(color)
-	var/list/HSL = rgb2hsl(hex2num(copytext(color,2,4)),hex2num(copytext(color,4,6)),hex2num(copytext(color,6,8)))
+	var/list/HSL = rgb2hsl(hex2num(copytext_char(color, 2, 4)), hex2num(copytext_char(color, 4, 6)), hex2num(copytext_char(color, 6, 8)))
```
Generally i feel all these color helpers would be better off with `if(length != length_char) CRASH` somewhere.

MrPerson replied:
I tried to see if there were any special Unicode symbols for hex digits and there does not appear to be any, which makes sense. I'll also assume # is always gonna be the opener. Thus yeah I agree with you.

MrPerson committed:
Assume color hex strings are always # followed by ascii.
Properly encodes and decodes the stuff in mob_helpers.dm which fixes some issues there.

MrPerson committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into unicode_horn_part_2a

# Conflicts:
#	code/game/objects/items/cards_ids.dm
#	code/modules/admin/verbs/maprotation.dm

SpaceManiac commented:
> This is still VERY rough btw; DNA is a mess, the status displays are complete ass, there's a copytext() in code\datums\shuttles.dm that I'm not sure what to do with, there's an issue with stuttering I need to figure out, and I didn't touch anything in the tools folder.

Is all this stuff still broken?

actioninja commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -51,134 +51,127 @@
-	return sanitize(t)
+			. += "*"
+	return sanitize(.)
+
```
not convinced this is going to work considering that index gets bytes

MrPerson replied:
You're gonna have to be more specific

MrStonedOne replied:
http://www.byond.com/forum/post/2520672#comment25078860
> To handle these special cases, BYOND has introduced _char versions of most of the text procs, like findtext_char() and copytext_char(). Here is one way you could, but shouldn't, update the previous code:
> 
> ```
> // This works in 513+ but performance sucks for long strings
> proc/AnimateMaptext(object, text)
>     var/i, len=length_char(text)
>     for(i=2, i<=len, ++i)
>         object.maptext = copytext_char(text, 1, i)
>         sleep(1)
> 
> ```
> Counting characters is not necessarily a fast process, so for long text strings this would be cumbersome. Here's a better way:
> 
> ```
> // This is better (and will also work in 512!)
> proc/AnimateMaptext(object, text)
>     var/i, len=length(text)
>     var/ch
>     for(i=1, i<=len,)
>         // the read-only [] operator will grab the whole character at byte i
>         i += length(text[i])
>         object.maptext = copytext(text, 1, i)
>         sleep(1)
> ```

Byond/lummox suggests that index reads are a better solution.

MrStonedOne replied:
The key piece you might be missing is that the length of char is added to i, ensuring it is aligned to the start of the next character.

actioninja commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -51,134 +51,127 @@
   */
-/proc/slur(n)
-	var/phrase = html_decode(n)
+/proc/slur(phrase)
```
same for this one

actioninja commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -51,134 +51,127 @@
-		if (prob(80) && (ckey(n_letter) in list("b","c","d","f","g","h","j","k","l","m","n","p","q","r","s","t","v","w","x","y","z")))
-			if (prob(10))
-				n_letter = text("[n_letter]-[n_letter]-[n_letter]-[n_letter]")//replaces the current letter with this instead.
+/proc/stutter(phrase)
```
same for this one

actioninja commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -204,15 +197,20 @@
   * text is the inputted message, replace_characters will cause original letters to be replaced and chance are the odds that a character gets modified.
   */
 /proc/Gibberish(text, replace_characters = FALSE, chance = 50)
```
same here

actioninja commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -224,28 +222,36 @@
+
+		. += stutterphrase
+		stutterphrase = ""
+	return sanitize(.)
```
this is actually unused, just kill it.

MrPerson replied:
Are you serious? Goddamnit I spent the most time on this one.

SpaceManiac commented on code/__HELPERS/icons.dm:
```diff
@@ -981,7 +981,7 @@ world
 
 	if(!letter)
-		letter = copytext(A.name, 1, 2)
+		letter = A.name[1]
```
x[1] starts at byte offset 1 and returns a single code point

MrPerson replied:
It'll return the entire character no matter how many bytes it is.

FireFlashie replied:
@MrPerson,
`The read-only [] index operator also uses byte positions.`
According to http://www.byond.com/docs/ref/index.html#/%7Bnotes%7D/Unicode

MrPerson committed:
Removes ninjaspeak since it's unused

MrPerson commented:
I fixed the stuttering issue. The shuttle copytext() should be ok as long as nobody uses multibyte characters in map tokens (the "aaa" "aab" bits).

Status displays are awful, but they're awful NOW so that doesn't really change anything. They'll break real hardcore if people start sticking in unrendered unicode symbols. We need helpers to tell how many "real" symbols long a string is to fix that problem though.
